src/ray/core_worker/object_interface.cc

```diff
@@ -139,92 +125,77 @@ Status CoreWorkerObjectInterface::Wait(const std::vector<ObjectID> &ids, int num
   (*results).resize(ids.size(), false);
 
   if (num_objects <= 0 || num_objects > static_cast<int>(ids.size())) {
-    return Status::Invalid("num_objects value is not valid");
+    return Status::Invalid(
+        "Number of objects to wait for must be between 1 and the number of ids.");
   }
 
   EnumUnorderedMap<StoreProviderType, std::unordered_set<ObjectID>>
       object_ids_per_store_provider;
   GroupObjectIdsByStoreProvider(ids, &object_ids_per_store_provider);
 
+  size_t total_count = 0;
+  for (const auto &entry : object_ids_per_store_provider) {
+    total_count += entry.second.size();
+  }
+  if (total_count != ids.size()) {
+    return Status::Invalid("Duplicated object IDs not supported in wait.");
+  }
+
+  // TODO(edoakes): this logic is not ideal, and will have to be addressed
+  // before we enable direct actor calls in the Python code. If we are waiting
+  // on a list of objects mixed between multiple store providers, we could
+  // easily end up in the situation where we're blocked waiting on one store
+  // provider while another actually has enough objects ready to fulfill
+  // 'num_objects'. This is partially addressed by trying them all once with
+  // a timeout of 0, but that does not address the situation where objects
+  // become available on the second store provider while waiting on the first.
+
+  std::unordered_set<ObjectID> ready;
   // Wait from all the store providers with timeout set to 0. This is to avoid the case
   // where we might use up the entire timeout on trying to get objects from one store
   // provider before even trying another (which might have all of the objects available).
-  RAY_RETURN_NOT_OK(WaitFromMultipleStoreProviders(ids, object_ids_per_store_provider,
-                                                   /* timeout_ms= */ 0, &num_objects,
-                                                   results));
+  RAY_RETURN_NOT_OK(WaitFromMultipleStoreProviders(object_ids_per_store_provider,
+                                                   /*timeout_ms=*/0, num_objects,
+                                                   &ready));
 
-  if (num_objects > 0) {
+  if (ready.size() < static_cast<size_t>(num_objects)) {
     // Wait from all the store providers with the specified timeout
     // if the required number of objects haven't been ready yet.
-    RAY_RETURN_NOT_OK(WaitFromMultipleStoreProviders(ids, object_ids_per_store_provider,
-                                                     /* timeout_ms= */ timeout_ms,
-                                                     &num_objects, results));
+    RAY_RETURN_NOT_OK(WaitFromMultipleStoreProviders(object_ids_per_store_provider,
+                                                     /*timeout_ms=*/timeout_ms,
+                                                     num_objects, &ready));
```

Comment:
seems like there is a bug here.

Wait() for each store provider would remove the available ones from its input object ids, which means it modifies `object_ids_per_store_provider`, so when `WaitFromMultipleStoreProviders` is called for the second time, all the object ids that are available at the first call of this function have already been removed from `object_ids_per_store_provider` , but this time it's still expecting `num_objects` objects to be available, which could block forever...

Reply:
Ach yeah, I think you're right. Shows that this logic is indeed too complex as-is. Note that there's another bug in the existing code because we don't remove ready object IDs from the ready set, so they can be double-counted by the `Wait()` call with timeout=0 and then the one with a timeout.

Reply:
yes.. the logic here is indeed a bit complex